File: src/main/java/org/polyfrost/fovtweaks/hooks/FOVHook.java
package org.polyfrost.fovtweaks.hooks;

import dev.deftu.omnicore.client.OmniClientPlayer;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.init.Items;
import net.minecraft.potion.Potion;
import net.minecraft.potion.PotionEffect;
import org.polyfrost.fovtweaks.FOVTweaks;

public class FOVHook {
    public static float getFov() {
        float base = 1.0f;
        EntityPlayer player = OmniClientPlayer.getInstance();

        if (player == null) {
            return base;
        }

        if (FOVTweaks.config.useFovModifiers) {
            base *= getFovModifier(player);
        }
        if (FOVTweaks.config.removeWaterFov) {
            base *= getWaterFovModifier(player);
        }
        return base;
    }

    private static float getFovModifier(EntityPlayer player) {
        float modifier = 1.0f;

        PotionEffect speed = player.getActivePotionEffect(Potion.moveSpeed);
        PotionEffect slowness = player.getActivePotionEffect(Potion.moveSlowdown);

        if (player.capabilities.isFlying) modifier += 0.1f * FOVTweaks.config.flyingModifier;

        if (player.isSprinting()) modifier += 0.15f * FOVTweaks.config.sprintingModifier;

        if (speed != null) modifier += 0.1f * (speed.getAmplifier() + 1f) * FOVTweaks.config.speedModifier;

        if (slowness != null) modifier += -0.075f * (slowness.getAmplifier() + 1f) * FOVTweaks.config.slownessModifier;

        if (player.capabilities.getWalkSpeed() == 0f || Float.isNaN(modifier) || Float.isInfinite(modifier))
            modifier = 1.0F;

        if (player.isUsingItem() && player.getItemInUse().getItem() == Items.bow) {
            int useDuration = player.getItemInUseDuration();
            float duration = useDuration / 20f;
            float f1 = duration > 1f ? 1f : duration * duration;

            modifier *= 1f - f1 * 0.15F * FOVTweaks.config.bowModifier;
        }

        return modifier;
    }

    private static float getWaterFovModifier(EntityPlayer player) {
        return player.isInWater() ? 70f / 60f : 1.0f;
    }
}

File: src/main/kotlin/org/polyfrost/fovtweaks/hooks/FOVHook.kt
package org.polyfrost.fovtweaks.hooks

import dev.deftu.omnicore.client.OmniClientPlayer.getInstance
import net.minecraft.client.entity.EntityPlayerSP
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.init.Items
import net.minecraft.potion.Potion
import org.polyfrost.fovtweaks.FOVTweaks

object FOVHook {

    private const val WATER_MULTIPLIER = 70f / 60f

    @JvmStatic
    fun getFov(): Float {
        var base = 1.0f
        val player: EntityPlayerSP = getInstance() ?: return base

        if (FOVTweaks.config.useFovModifiers) {
            base *= getFovModifier(player)
        }
        if (FOVTweaks.config.removeWaterFov) {
            base *= getWaterFovModifier(player)
        }
        return base
    }

    private fun getFovModifier(player: EntityPlayer): Float {
        var modifier = 1.0f

        val speed = player.getActivePotionEffect(Potion.moveSpeed)
        val slowness = player.getActivePotionEffect(Potion.moveSlowdown)

        if (player.capabilities.isFlying) modifier += 0.1f * FOVTweaks.config.flyingModifier

        if (player.isSprinting) modifier += 0.15f * FOVTweaks.config.sprintingModifier

        if (speed != null) modifier += 0.1f * (speed.amplifier + 1f) * FOVTweaks.config.speedModifier

        if (slowness != null) modifier += -0.075f * (slowness.amplifier + 1f) * FOVTweaks.config.slownessModifier

        if (player.capabilities.walkSpeed == 0f || modifier.isNaN() || modifier.isInfinite()) modifier = 1.0f

        if (player.isUsingItem && player.itemInUse.item === Items.bow) {
            val useDuration = player.itemInUseDuration
            val duration = useDuration / 20f
            val f1 = if (duration > 1f) 1f else duration * duration

            modifier *= 1f - f1 * 0.15f * FOVTweaks.config.bowModifier
        }

        return modifier
    }

    private fun getWaterFovModifier(player: EntityPlayer) = WATER_MULTIPLIER.takeIf { player.isInWater } ?: 1.0f
}